Invalidate scope-breaking :has(:is(...)) selectors more thoroughly

heycam opened the pull request:
#### 3724105055b3d155fee1da5604623c6200cf1f28
<pre>
Invalidate scope-breaking :has(:is(...)) selectors more thoroughly
<a href="https://bugs.webkit.org/show_bug.cgi?id=253944">https://bugs.webkit.org/show_bug.cgi?id=253944</a>
rdar://106768250

Reviewed by Antti Koivisto.

It&apos;s possible to write :has() selectors with arguments that can match
elements outside of the :has scope, such as

:has(:is(.x .y))

which could match the .x against an ancestor of the :has scope, and

:has(~ :is(.x .y))

which could match the .x against an ancestor of a later sibling of the
:has scope.

We currently handle scope breaking :has() selectors by generating
RuleFeatures based on the nested selectors&apos; features and storing them
with MatchElement::HasNonSubjectOrScopeBreaking, which causes
invalidation to match the inner scope-breaking selector (the :is())
against all elements in the document.

Because we store these in InvalidationRuleSets keyed off the nested
selectors&apos; features, it means that for the above examples, we will
process these document-wide invalidations whenever a class name changes
to or from x or y. But this is not sufficient to invalidate for
selectors like

:has(:is(.x + .y .z))

where if the .x and .y elements are outside the :has scope, an element
insertion or removal between them will not find any relevant
invalidation rule sets, and

:has(~ :is(.x ~ .y))

for similar reasons.

This PR changes how we collect scope breaking :has() rules, so that in
addition to generating entries in m_hasPseudoClassInvalidationRuleSets
on Style::ScopeRuleSet, we also record them in a new
m_scopeBreakingHasPseudoClassInvalidationRuleSet, which we look up for
every element insertion or removal. For simplicity, we do this more
drastic invalidation for both of the above cases.

MatchElement::HasNonSubjectOrScopeBreaking is split into two separate
values -- HasNonSubject and HasScopeBreaking -- where only the latter
causes an entry to be added to
m_scopeBreakingHasPseudoClassInvalidationRuleSet.

This change does not regress the performance of invalidation of :has()
selectors that do not contain logical combination pseudo-classes like
:is() and :not(), or which have do have one but do not have any selector
combinators inside them.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/is-pseudo-containing-complex-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/not-pseudo-containing-complex-in-has-expected.txt:
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
(WebCore::Style::ChildChangeInvalidation::invalidateForChangeOutsideHasScope):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasBeforeMutation):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasAfterMutation):
(WebCore::Style::needsDescendantTraversal):
* Source/WebCore/style/ChildChangeInvalidation.h:
* Source/WebCore/style/ClassChangeInvalidation.cpp:
(WebCore::Style::ClassChangeInvalidation::computeInvalidation):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::isSiblingOrSubject):
(WebCore::Style::isHasPseudoClassMatchElement):
(WebCore::Style::isScopeBreaking):
(WebCore::Style::computeNextHasPseudoClassMatchElement):
(WebCore::Style::computeHasPseudoClassMatchElement):
(WebCore::Style::computeSubSelectorMatchElement):
(WebCore::Style::RuleFeatureSet::recursivelyCollectFeaturesFromSelector):
(WebCore::Style::RuleFeatureSet::collectFeatures):
(WebCore::Style::RuleFeatureSet::add):
(WebCore::Style::RuleFeatureSet::clear):
(WebCore::Style::RuleFeatureSet::shrinkToFit):
* Source/WebCore/style/RuleFeature.h:
(WebCore::Style::RuleFeatureSet::usesMatchElement const):
(WebCore::Style::RuleFeatureSet::usesHasPseudoClass const):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateStyleWithMatchElement):
(WebCore::Style::Invalidator::invalidateWithScopeBreakingHasPseudoClassRuleSet):
* Source/WebCore/style/StyleInvalidator.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::collectFeatures const):
* Source/WebCore/style/StyleScopeRuleSets.h:
(WebCore::Style::ScopeRuleSets::scopeBreakingHasPseudoClassInvalidationRuleSet const):

Canonical link: <a href="https://commits.webkit.org/268038@main">https://commits.webkit.org/268038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c68cad6d53ee7a0e574725917c4963ad52e4ee2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18704 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20212 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17193 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22008 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18852 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19113 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18590 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18786 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15999 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21092 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16009 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16756 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23238 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16926 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21125 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14846 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16579 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20943 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2274 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/RuleFeature.cpp:
```diff
@@ -79,6 +79,34 @@ bool isHasPseudoClassMatchElement(MatchElement matchElement)
+    switch (matchElement) {
+    case MatchElement::HasAnySibling:
+    case MatchElement::HasNonSubjectOrScopeBreaking:
+        return true;
```
We should split HasNonSubjectOrScopeBreaking into two separate cases. HasNonSubject does not need scope breaking handling.

heycam committed:
Invalidate scope-breaking :has(:is(...)) selectors more thoroughly
https://bugs.webkit.org/show_bug.cgi?id=253944
rdar://106768250

Reviewed by Antti Koivisto.

It's possible to write :has() selectors with arguments that can match
elements outside of the :has scope, such as

:has(:is(.x .y))

which could match the .x against an ancestor of the :has scope, and

:has(~ :is(.x .y))

which could match the .x against an ancestor of a later sibling of the
:has scope.

We currently handle scope breaking :has() selectors by generating
RuleFeatures based on the nested selectors' features and storing them
with MatchElement::HasNonSubjectOrScopeBreaking, which causes
invalidation to match the inner scope-breaking selector (the :is())
against all elements in the document.

Because we store these in InvalidationRuleSets keyed off the nested
selectors' features, it means that for the above examples, we will
process these document-wide invalidations whenever a class name changes
to or from x or y. But this is not sufficient to invalidate for
selectors like

:has(:is(.x + .y .z))

where if the .x and .y elements are outside the :has scope, an element
insertion or removal between them will not find any relevant
invalidation rule sets, and

:has(~ :is(.x ~ .y))

for similar reasons.

This PR changes how we collect scope breaking :has() rules, so that in
addition to generating entries in m_hasPseudoClassInvalidationRuleSets
on Style::ScopeRuleSet, we also record them in a new
m_scopeBreakingHasPseudoClassInvalidationRuleSet, which we look up for
every element insertion or removal. For simplicity, we do this more
drastic invalidation for both of the above cases.

MatchElement::HasNonSubjectOrScopeBreaking is split into two separate
values -- HasNonSubject and HasScopeBreaking -- where only the latter
causes an entry to be added to
m_scopeBreakingHasPseudoClassInvalidationRuleSet.

This change does not regress the performance of invalidation of :has()
selectors that do not contain logical combination pseudo-classes like
:is() and :not(), or which have do have one but do not have any selector
combinators inside them.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/is-pseudo-containing-complex-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/not-pseudo-containing-complex-in-has-expected.txt:
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
(WebCore::Style::ChildChangeInvalidation::invalidateForChangeOutsideHasScope):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasBeforeMutation):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasAfterMutation):
(WebCore::Style::needsDescendantTraversal):
* Source/WebCore/style/ChildChangeInvalidation.h:
* Source/WebCore/style/ClassChangeInvalidation.cpp:
(WebCore::Style::ClassChangeInvalidation::computeInvalidation):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::isSiblingOrSubject):
(WebCore::Style::isHasPseudoClassMatchElement):
(WebCore::Style::isScopeBreaking):
(WebCore::Style::computeNextHasPseudoClassMatchElement):
(WebCore::Style::computeHasPseudoClassMatchElement):
(WebCore::Style::computeSubSelectorMatchElement):
(WebCore::Style::RuleFeatureSet::recursivelyCollectFeaturesFromSelector):
(WebCore::Style::RuleFeatureSet::collectFeatures):
(WebCore::Style::RuleFeatureSet::add):
(WebCore::Style::RuleFeatureSet::clear):
(WebCore::Style::RuleFeatureSet::shrinkToFit):
* Source/WebCore/style/RuleFeature.h:
(WebCore::Style::RuleFeatureSet::usesMatchElement const):
(WebCore::Style::RuleFeatureSet::usesHasPseudoClass const):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateStyleWithMatchElement):
(WebCore::Style::Invalidator::invalidateWithScopeBreakingHasPseudoClassRuleSet):
* Source/WebCore/style/StyleInvalidator.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::collectFeatures const):
* Source/WebCore/style/StyleScopeRuleSets.h:
(WebCore::Style::ScopeRuleSets::scopeBreakingHasPseudoClassInvalidationRuleSet const):

Canonical link: https://commits.webkit.org/268038@main
